File: src/main/java/io/dipcoin/sui/perp/model/ApiResponse.java
/*
 * Copyright 2025 Dipcoin LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");you may not use this file except in compliance with
 * the License.You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,software distributed under the License is distributed on
 * an "AS IS" BASIS,WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.See the License for the
 * specific language governing permissions and limitations under the License.
 */

package io.dipcoin.sui.perp.model;

import io.dipcoin.sui.perp.exception.ErrorCode;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * @author : Same
 * @datetime : 2025/10/21 10:59
 * @Description :
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
public class ApiResponse<T> {

    private int code;

    private String message;

    private T data;

    public static <T> ApiResponse<T> success(T data) {
        ApiResponse<T> response = new ApiResponse<>();
        response.setCode(ErrorCode.SUCCESS.getCode());
        response.setMessage(ErrorCode.SUCCESS.getDesc());
        response.setData(data);
        return response;
    }

    public static <T> ApiResponse<T> success() {
        ApiResponse<T> response = new ApiResponse<>();
        response.setCode(ErrorCode.SUCCESS.getCode());
        response.setMessage(ErrorCode.SUCCESS.getDesc());
        return response;
    }

    public static <T> ApiResponse<T> error(String message) {
        ApiResponse<T> response = new ApiResponse<>();
        response.setCode(ErrorCode.SYSTEM_ERROR.getCode());
        response.setMessage(message);
        return response;
    }

    public static <T> ApiResponse<T> error(int code, String message) {
        ApiResponse<T> response = new ApiResponse<>();
        response.setCode(code);
        response.setMessage(message);
        return response;
    }

    public static <T> ApiResponse<T> error(ErrorCode errorCode) {
        ApiResponse<T> response = new ApiResponse<>();
        response.setCode(errorCode.getCode());
        response.setMessage(errorCode.getDesc());
        return response;
    }
}

File: src/main/kotlin/io/dipcoin/sui/perp/model/ApiResponse.kt
/*
 * Copyright 2025 Dipcoin LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");you may not use this file except in compliance with
 * the License.You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,software distributed under the License is distributed on
 * an "AS IS" BASIS,WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.See the License for the
 * specific language governing permissions and limitations under the License.
 */

package io.dipcoin.sui.perp.model

import io.dipcoin.sui.perp.exception.ErrorCode

data class ApiResponse<T>(
    var code: Int = 0,
    var message: String? = null,
    var timezone: Long? = null,
    var data: T? = null,
) {
    companion object {
        fun <T> success(data: T): ApiResponse<T> = ApiResponse(
            code = ErrorCode.SUCCESS.code,
            message = ErrorCode.SUCCESS.desc,
            data = data
        )

        fun <T> success(): ApiResponse<T> = ApiResponse(
            code = ErrorCode.SUCCESS.code,
            message = ErrorCode.SUCCESS.desc
        )

        fun <T> error(message: String): ApiResponse<T> = ApiResponse(
            code = ErrorCode.SYSTEM_ERROR.code,
            message = message
        )

        fun <T> error(code: Int, message: String): ApiResponse<T> = ApiResponse(
            code = code,
            message = message
        )

        fun <T> error(errorCode: ErrorCode): ApiResponse<T> = ApiResponse(
            code = errorCode.code,
            message = errorCode.desc
        )
    }
}
